### DISCUSSION TITLE
Plejd scene call as trigger in HA

### DISCUSSION THREAD

**Discussion by n0rdstrand** [icanos/hassio-plejd#198] [Q&A]
Hey, im very new to all of this but after alot of reading here i am a little confused if i am or am not able to use my plejd scene calls as triggers in HA for other HA scenarios. I have the latest dev version installed with the plejd scenes actually being scenes in HA. But i cant get it to work. 
This is what i get in the log when activating the scene from my physical plejd button.

2021-05-19 17:21:47 VRB [plejd-ble] Raw event received: 02011000210700000080
2021-05-19 17:21:47 VRB [plejd-ble] Decoded: Device null (BLE address 2), cmd 21, state 7, dim 0
2021-05-19 17:21:47 DBG [plejd-ble] Plejd AV (7) scene triggered (device id null).
2021-05-19 17:21:47 VRB [plejd-mqtt] Scene triggered: 4df35c58-67c7-476f-a3a4-3c2679c14de3
2021-05-19 17:21:47 VRB [plejd-mqtt] Received mqtt message on homeassistant/device_automation/plejd/4df35c58-67c7-476f-a3a4-3c2679c14de3_trigger/state
2021-05-19 17:21:47 VRB [plejd-mqtt] Getting device 4df35c58-67c7-476f-a3a4-3c2679c14de3_trigger from registry
2021-05-19 17:21:47 VRB [plejd-mqtt] Sent mqtt state command for device_automation,  (4df35c58-67c7-476f-a3a4-3c2679c14de3_trigger).

<img width="753" alt="plejd" src="https://user-images.githubusercontent.com/84462198/118839933-3b6e0200-b8c7-11eb-8048-5ac272a8c328.png">
<img width="752" alt="plejd 2" src="https://user-images.githubusercontent.com/84462198/118840131-65bfbf80-b8c7-11eb-9fab-784ee1bfb790.png">

**Comment by SweVictor** [comment]
The "scene triggered" data provided in the dev system (0.8.0-dev) is based on Home Assistant "device automation" over MQTT. https://www.home-assistant.io/integrations/device_trigger.mqtt/

I have honestly not configured automation based on this, so I'm not sure exactly how to pick them up in HA. Either way I think it's something we need to address in the readme before we release.

@faanskit I think has played around a bit more with this (and WPH-01 which uses the same type), maybe you can help?

**Comment by faanskit** [comment]
I cannot get scene triggers to work either. 

WPH-01 device_triggers works fine, but something is weird with the device_triggers for the scenes.
I'll do some debugging. On surface it looks fine...

Difference:

**WPH-01**
```
payload: '1'
qos: 1
topic: homeassistant/device_automation/plejd/F7EFCACF2614/state
type: button_short_press
```
**Scene**
```
type: scene
subtype: trigger
```

Payload should really be optional.

**Comment by faanskit** [comment]
**Test 1** 
Replicate manual steps to replicate WPH-01 behavior, with similar data as with Scene
Status: **OK**

```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman_1/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/batman","payload":"1","qos":1,"topic":"~/state","type":"button_short_press","subtype":"button_1","device":{"identifiers":"batman","manufacturer":"Plejd","model":"TEST","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman_0/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/batman","payload":"0","qos":1,"topic":"~/state","type":"button_short_press","subtype":"button_2","device":{"identifiers":"batman","manufacturer":"Plejd","model":"TEST","name":"Batman"}}'

mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman/state' -m '1'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman/state' -m '0'
```

**Test 2** 
Change from `button_short_press`/`button_1 `to `scene`/`trigger` (and  `scene`/`trigger_2`)
Status: **OK**

```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman_1/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/batman","payload":"1","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"batman","manufacturer":"Plejd","model":"TEST","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman_0/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/batman","payload":"0","qos":1,"topic":"~/state","type":"scene","subtype":"trigger_2","device":{"identifiers":"batman","manufacturer":"Plejd","model":"TEST","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman/state' -m '1'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman/state' -m '0'
```

**Test 3** 
Remove payload to fully mimic Scene device_trigger, without having an actual scene.

Status: **OK**

```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman_1/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/batman","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"batman","manufacturer":"Plejd","model":"TEST","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman/state' -m ''
```


**Hypothesis:**
Home Assistant get confused when a Scene and a Trigger is used at the same time.
More debugging later...

```
homeassistant/scene/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603/config 
{"name":"LeftOn",
"unique_id":"ed815a49-5f12-4def-bb90-edc90e5c2603",
"~":"homeassistant/scene/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603",
"command_topic":"~/set",
"availability_topic":"~/availability",
"payload_on":"ON",
"qos":1,
"retain":false}
```

```
homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/config 
{"automation_type":"trigger",
"~":"homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603",
"qos":1,
"topic":"~/state",
"type":"scene",
"subtype":"trigger",
"device":{"identifiers":"ed815a49-5f12-4def-bb90-edc90e5c2603",
"manufacturer":"Plejd",
"model":"Scene",
"name":"LeftOn"}}
```

**Comment by faanskit** [comment]
**TEST 6** 
From a working config, add more details. Added trigger topic.
Status: **OK**
```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/batman_1/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","manufacturer":"Plejd","model":"Scene","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/state' -m ''
```
**TEST 7**
From a working config, add more details. Added config topic.
For some reason, the config topic fails.
Status: **FAIL**
```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","manufacturer":"Plejd","model":"Scene","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/state' -m ''
```
**TEST 11**
Removed one character (first 'e') from config topic
Status: **OK**
```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/d815a49-5f12-4def-bb90-edc90e5c2603_trigger/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","manufacturer":"Plejd","model":"Scene","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/state' -m ''
```
**TEST 12**
Removed one character (last 'r') from config topic
Status: **OK**
```
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigge/config' -m '{"automation_type":"trigger","~":"homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"ed815a49-5f12-4def-bb90-edc90e5c2603_trigger","manufacturer":"Plejd","model":"Scene","name":"Batman"}}'
mosquitto_pub -h 192.168.1.192 -u homeassistant -P Leey1Eibae6kai3ooThae4aeP6quu8ahRei6oirooh2Ais8hee6pheizaijee4Wo -t 'homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/state' -m ''
```

### Hypothesis
Home Assistant cannot handle too long MQTT topics.
**FAIL**
`homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigger/config`
**OK**
`homeassistant/device_automation/plejd/ed815a49-5f12-4def-bb90-edc90e5c2603_trigge/config`
`homeassistant/device_automation/plejd/d815a49-5f12-4def-bb90-edc90e5c2603_trigger/config`

**Comment by faanskit** [comment]
Quick code change: `trigger` ==> `trig`
**MqttClient.js**
`const getTriggerUniqueId = (/** @type { string } */ uniqueId) => `${uniqueId}_trig`;
`
Still problematic. No device_triggers appears for scenes.

Second issue seems related to naming of device triggers. With hands on changes to name; looks to be working.
Can a device_trigger and a scene have the same name?!?

**Comment by faanskit** [comment]
### MqttClient.js
Second change, can the name be what is causing this?!?
**getInputDeviceTriggerDiscoveryPayload**

```
name: sceneDevice.name` ==> `name: `${sceneDevice.name}_Trigger`
```

**Nope**, but unique_id from Scene and identifiers from Device_trigger may be the problem

```
homeassistant/scene/plejd/8d74b717-9fe4-496a-b708-6d38786bc506/config {"name":"LeftOff","unique_id":"8d74b717-9fe4-496a-b708-6d38786bc506","~":"homeassistant/scene/plejd/8d74b717-9fe4-496a-b708-6d38786bc506","command_topic":"~/set","availability_topic":"~/availability","payload_on":"ON","qos":1,"retain":false}
homeassistant/device_automation/plejd/8d74b717-9fe4-496a-b708-6d38786bc506_trig/config {"automation_type":"trigger","~":"homeassistant/device_automation/plejd/8d74b717-9fe4-496a-b708-6d38786bc506","qos":1,"topic":"~/state","type":"scene","subtype":"trigger","device":{"identifiers":"8d74b717-9fe4-496a-b708-6d38786bc506","manufacturer":"Plejd","model":"Scene","name":"LeftOff_Trigger"}}
```

### MqttClient.js
Third change, can the unique_id  be what is causing this?!?
**getInputDeviceTriggerDiscoveryPayload**

```
    identifiers: `${sceneDevice.uniqueId}`, == >  identifiers: `${sceneDevice.uniqueId}_trigger`,
```

### MqttClient.js
Fourth change, and now my own little bug
**getInputDeviceTriggerDiscoveryPayload**

```
  '~': getBaseTopic(sceneDevice.uniqueId, MQTT_TYPES.DEVICE_AUTOMATION),
==>
  '~': getBaseTopic(`${sceneDevice.uniqueId}_trig`, MQTT_TYPES.DEVICE_AUTOMATION),

```


### Total solution
**Change 1: MqttClient.js**
```
const getTriggerUniqueId = (/** @type { string } */ uniqueId) => `${uniqueId}_trig`;
```

**Change 2: MqttClient.js**
```
const getSceneDeviceTriggerhDiscoveryPayload = (
  /** @type {import('./types/DeviceRegistry').OutputDevice} */ sceneDevice,
) => ({
  automation_type: 'trigger',
  '~': getBaseTopic(`${sceneDevice.uniqueId}_trig`, MQTT_TYPES.DEVICE_AUTOMATION),
  qos: 1,
  topic: `~/${TOPIC_TYPES.STATE}`,
  type: 'scene',
  subtype: 'trigger',
  device: {
    identifiers: `${sceneDevice.uniqueId}_trigger`,
    manufacturer: 'Plejd',
    model: sceneDevice.typeName,
    name: sceneDevice.name,
  },
});
```

**Comment by faanskit** [comment]
### Conclusion
There were two issues at play here.

**Issue 1** - Home Assistant fails to use topics > 90 characters. 
_Solution_: Reduce topic length for device trigger (workaround?)

**Issue 2** - MQTT Scenes and MQTT Device_triggers gets confused when scene unique id is used for device_trigger[device[identifiers]]
_Solution_: Add suffix to identifier for the device_trigger to make it unique

### Solution:
Three lines of code to change in **MqttClient.js**
38: change **_trigger** to **_trig**
113: change **sceneDevice.uniqueId** to **\`${sceneDevice.uniqueId}_trig\`**
119: change  **\`${sceneDevice.uniqueId}\`** to **\`${sceneDevice.uniqueId}_trigger\`**

**Comment by faanskit** [comment]
### Morning update
I did not want to believe that the topic length was an issue, so I removed changes in line 38 and 113, assuming this was only a _unique-id_ problem.

Sadly the test failed, and it really messes with Home Assistants device registry. Even when reverting the changes, Ui behaves badly and the failed MQTT devices does not show up (and cannot be removed).

Some manual restoration and the system is back up with the all three lines of code changed.


### Noted behavior
When you trigger a  Plejd scene from Plejd App, or from a Plejd Device, this is picked up as a device_trigger in Home Assistant.
**But:**
When you trigger the Plejd scene from Home Assistant Scene, the device_trigger is not seen. This is a feature, not a bug.

One can argue that even when Plejd scenes are performed by Home Assistant, the device_trigger shall be triggered to have consistent behavior.

**Discuss...**

**Reply by faanskit** [reply]
PR #199 should solve this issue. Maybe not the best design, but @SweVictor will give his remarks before it is merged.
Still the above discussion is needed...

**Reply by SweVictor** [reply]
Very strange the topic length issue, but great job finding the core issue!

I agree that scenes triggered from HA should trigger the device trigger as well, so I'd consider that ... a missing feature 😄
